### PR TITLE
feat: Update logits bitmask kernel to v3

### PR DESCRIPTION
**Pull request by syuoni**
The XGrammar team provides important insights on the kernel workload. In most cases, the bitmask tensor is almost-full (bit values are 1) and almost-empty (bit values are 0).

Compared the kernel version on main (v2), the PR introduces the kernel developed in https://github.com/mlc-ai/xgrammar/pull/186 (v3):
* The kernel v3 shows ~1.3x and ~2.0x speedup on large batch sizes for the almost-full and almost-empty scenarios, respectively.
* The kernel v3 slightly sacrifices the performance on half-full scenario, compared to v2.

See https://github.com/mlc-ai/xgrammar/tree/main/examples/benchmark#benchmark-apply-token-bitmask-inplace-kernels for more perf numbers. Please see https://github.com/mlc-ai/xgrammar/pull/186 for more background.

### CODE REVIEW

**Comment by niukuo** [comment]
[PR_Github #253](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/253/) [ run ] triggered by Bot

**Comment by niukuo** [comment]
[PR_Github #292](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/292/) [ run ] triggered by Bot

**Comment by niukuo** [comment]
[PR_Github #253](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/253/) [ run ] completed with state `ABORTED`
[/LLM/main/L0_MergeRequest_PR pipeline #248](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/248/) completed with status: 'FAILURE'

**Comment by niukuo** [comment]
[PR_Github #292](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/292/) [ run ] completed with state `FAILURE`
[/LLM/main/L0_MergeRequest_PR pipeline #281](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/281/) completed with status: 'FAILURE'

**Comment by niukuo** [comment]
[PR_Github #310](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/310/) [ run ] triggered by Bot

**Comment by niukuo** [comment]
[PR_Github #310](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/310/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #294](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/294/) completed with status: 'FAILURE'

**Comment by niukuo** [comment]
[PR_Github #347](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/347/) [ run ] triggered by Bot

**Comment by niukuo** [comment]
[PR_Github #347](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/347/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #319](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/319/) completed with status: 'FAILURE'

File: cpp/tensorrt_llm/kernels/logitsBitmask.cu [hunk 1]
```diff
@@ -49,57 +49,100 @@ __device__ __nv_bfloat16 negativeInfinity<__nv_bfloat16>()
...
-    {
-        int offset = (i + threadIdx.x) % warpSize;
-        if (bitmaskIdx * kBitsPerMaskElement + offset >= vocabSizePadded)
+        if (bitmaskVal == 0)
         {
             continue;
         }
-        if (!((bitmaskVal >> offset) & 1))
+
+        if (bitmaskVal == kPackedMask)
```
Comment (byshiue) [inline cpp/tensorrt_llm/kernels/logitsBitmask.cu:97]: Currently (case 1), you 
1. setup -inf to logitsGmemPtr (write global memory)
2. Store logitsGmemPtr to logitsReg (read global memory and write to shared memory)
3. set -inf to logitsReg (write shared memory)
4. Store logitsReg back to logitsGmemPtr (read shared memory and write to global memory)

So, there are 2 global write, 1 global read, 2 shared write and 1 shared read

Is it possible to change the order to the following (case 2)
1. Store logitsGmemPtr to logitsReg (read global memory and write to shared memory)
2. setup -inf to logitsReg (same index as step 1 in case 1)
3. setup -inf to logitsReg (same index as step 3 in case 1)
4. store logitsReg back to logitsGmemPtr (read shared memory and write to global memory)

Comment (syuoni) [reply cpp/tensorrt_llm/kernels/logitsBitmask.cu:97]: Thanks for the review. I'm not sure if I fully understand your comments. This PR removes the use of shared memory; see the original `logitsSmem` has been removed.

With this PR, the kernel executes as:

* If `bitmaskVal` are all 0, the kernel skips all global memory access.
* If `bitmaskVal` are all 1, the kernel skips global memory load, and writes all `-inf` to global memory (1 global write).
* If `bitmaskVal` has both 0 and 1, the kernel loads global memory to registers `logitsReg`, writes `-inf` conditionally on the `bitmaskVal` values to `logitsReg`, and then writes `logitsReg` to global memory (1 global read and 1 global write).

Compared to previous kernel (the one on current main branch), this new kernel could skip a lot of memory accesses based on the value of `bitmaskVal`. That's where the acceleration comes from.

**Comment by niukuo** [comment]
[PR_Github #433](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/433/) [ run ] triggered by Bot

**Comment by niukuo** [comment]
[PR_Github #433](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/433/) [ run ] completed with state `FAILURE`
[/LLM/main/L0_MergeRequest_PR pipeline #371](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/371/) completed with status: 'FAILURE'

**Comment by niukuo** [comment]
[PR_Github #442](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/442/) [ run ] triggered by Bot

**Comment by niukuo** [comment]
[PR_Github #442](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/442/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #378](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/378/) completed with status: 'FAILURE'

**Comment by niukuo** [comment]
[PR_Github #491](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/491/) [ run ] triggered by Bot

**Comment by niukuo** [comment]
[PR_Github #491](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/491/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #423](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/423/) completed with status: 'SUCCESS'

**Comment by niukuo** [comment]
[PR_Github #527](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/527/) [ reuse-pipeline ] triggered by Bot

**Comment by niukuo** [comment]
[PR_Github #527](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/527/) [ reuse-pipeline ] completed with state `SUCCESS`
Reusing [PR_Github #491](https://prod.blsm.nvidia.com/sw-tensorrt-top-1/job/LLM/job/helpers/job/PR_Github/491/) for commit `60fd55d`

**Review by byshiue — APPROVED** [review]
LGTM

**Comment by niukuo** [comment]
[PR_Github #535](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/535/) [ reuse-pipeline ] triggered by Bot

**Comment by niukuo** [comment]
[PR_Github #535](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/535/) [ reuse-pipeline ] completed with state `SUCCESS`
Reusing [PR_Github #491](https://prod.blsm.nvidia.com/sw-tensorrt-top-1/job/LLM/job/helpers/job/PR_Github/491/) for commit `ff297de`
